keps/sig-node/20190530-commit-class.md

```diff
@@ -0,0 +1,405 @@
+I have a powerful tool in place to manage server footprint: the hypervisor
+overcommit settings. However, choosing the appropriate hypervisor commit level
+for a Kubernetes cluster is very challenging, because any overcommit level
+applies equally to user workloads and system- or kube- reserved resource
+slices. As a result, attempting to get high density at the VM level can result
+in unstable Kubernetes nodes.
+
+A more Kubernetes-aware commit level would let me isolate overcommit risk to
+user workloads: with `CommitClass` available, I could reduce (or eliminate) VM
+hypervisor overcommit.
+
+#### Kubernetes on VMs, no control over hypervisor / managed Kubernetes (cloud profile)
+
+As an operator of Kubernetes on VMs in the cloud, or of managed Kubernetes,
+I can use Cluster Autoscaling to ensure that my clusters are not provisioned
+wildly out of proportion to Pod resource requests. However, my toolset for
+achieving good utilization of my cluster is limited. A significant gap between
+Pod requests and Pod utilization aross the fleet will result in excessive
+cluster cost, as in other environments.
+
+Quoting from GKE's cluster autoscaling documentation:
+
+> Cluster autoscaler works based on **Pod resource requests**, based on how many
+> resources your Pods have *requested*. Cluster autoscaler does not take into
+> account the resources your Pods are actively *using*. Essentially, cluster
+> autoscaler trusts that the Pod resource requests you've provided are accurate
+> and schedules Pods on nodes based on that assumption.
+
+CommitClass would allow me to manage resource footprint in response to
+utilization, even as Cluster Autoscaler manages resource footprint in response
+to allocation.
+
+### Implementation Details
+
+#### Questions
+
+##### Conflict between multiple CommitClasses
+
+How do you resolve conflict if multiple nodes are selected by the same
+`CommitClass`? kubelet probably needs to record which `CommitClass` is in
+effect on a Node object so that it is clear to operators what's up. One
+alternative strategy would be to have a 'Binding' style object which enumerated
+the Node names it applied to, but that interacts badly with Cluster autoscaler.
+
+Nodes should default to a `CommitClass` where all percentages are '100'.
+
+##### Container CPU requests vs host physical CPUs
+
+It is usually a bad idea to give a single VM more vCPUs than physical CPUs are
+available on the host system. Should that be reflected here somehow for
+containers?
+
+##### Limiting supported resource types
+
+The proposed implementation limits the resources which can be scaled to
+**cpu**, **memory**, and **ephemeral-storage**. This is not a technical
+limitation: any of the resources enumerated in the
+`node.status.capacity|allocatable` objects could be scaled. Should the resource
+types be limited to a 'sane' subset?
+
+##### kube- and system-reserved resources
+
+Reserved resource slices should not be subject to `CommitClass` percentages, so
```

Comment:
This seems to make the problem a WHOLE LOT more complicated, doesn't it?

Reply:
I'm not sure. Specifically in terms of implementation in my hackish prototype, it was not so bad -- applying the scale factors after reserved resource chunks are subtracted from allocatable was enough to accomplish it. 

I might be too close to this to tell, but I also think that it isn't too complex conceptually: overcommiting my CPUs has the known risk of user workload performance degradation, but I think it's reasonable to expect it not to compromise system or kube components. I see this as a fairly compelling advantage over a VM hypervisor level approach to overcommit. 

Judging from your response, I feel like I must have a major blind spot, though :)

Reply:
I think what @thockin is getting at is that because resources are fungible, there's no way to cordon off "some" CPU for use by the kube-system or host that would **not** be impacted by workloads that consumed CPU resources on a node that has overcommit. In other words, there isn't a way to isolate "dedicated" CPU resources for the host or kube-system from impact by "floating" CPU resources.

The only way to do this kind of thing is to have separate pools of CPU resources: a pinset representing host logical processors that may take "shared" or "floating" CPUs and a pinset representing host logical processors that will service "dedicated" CPUs (for the host or for application containers). And that's a whole ball of ugly -- just warning ya :)

Reply:
What is "reserved"?  what heuristic do you use to know which things need special-casing (e.g. if kube-proxy runs as a daemonset - is it special?)

Reply:
Hmm. Yes. I was thinking exclusively about the kube- and system-reserved flags documented here: https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/.

However, kube-proxy as a daemonset is a great point, and touches on what you mentioned above about node-wide commit settings being too coarse. Perhaps it makes sense to try to think of a strategy that follows Priority Classes. In that approach, kube-proxy would be a 'critical' class pod, and operators would be unwise to overcommit resources for critical class pods. This would also align with decisions about admission/eviction, which respect priority class.